distsqlrun: fix range router test

madelynnblue opened the pull request:
There was a bug in the range router test because when an OID or INT
was the first column, they could create data that was after the last
span's end key, and thus trigger the DefaultDest bucket of 0. Fix
this by having the spans cover the entire input range.

While here, add sanity checks and tests to ensure incoming spans are
well-formed. This shouldn't ever be triggered with the current code,
but makes the range router a bit more future proof if someone adds
a new user or refactors how CSV works.

Fixes #17721
Fixes #17716

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17830)
<!-- Reviewable:end -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/distsqlrun/routers_test.go, line 267 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17830#-KsEcE3hvYv9E7ax-8Gv:-KsEcE3hvYv9E7ax-8Gw:badn6by) ([raw file](https://github.com/cockroachdb/cockroach/blob/27d7772e3882f579fbac71e983f7d853863d2441/pkg/sql/distsqlrun/routers_test.go#L267)):*
> ```Go
> 			{
> 				Start: []byte{testRangeRouterSpanBreak},
> 				End:   []byte{0xff},
> ```

can we use `keys.PrefixEnd(IntMax)` or one of those functions?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17830#-:-KsEcsWKBNFhfKvQO3Jt:bnfp4nl)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
distsqlrun: fix range router test

There was a bug in the range router test because when an OID or INT
was the first column, they could create data that was after the last
span's end key, and thus trigger the DefaultDest bucket of 0. Fix
this by having the spans cover the entire input range.

While here, add sanity checks and tests to ensure incoming spans are
well-formed. This shouldn't ever be triggered with the current code,
but makes the range router a bit more future proof if someone adds
a new user or refactors how CSV works.

Fixes #17721
Fixes #17716

mjibson commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/sql/distsqlrun/routers_test.go, line 267 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17830#-KsEcE3hvYv9E7ax-8Gv:-KsF8Do1gs-elK9lbhYc:bntinmg) ([raw file](https://github.com/cockroachdb/cockroach/blob/27d7772e3882f579fbac71e983f7d853863d2441/pkg/sql/distsqlrun/routers_test.go#L267)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

can we use `keys.PrefixEnd(IntMax)` or one of those functions?
</blockquote></details>

How about keys.MaxKey. That guarantees that nothing will be encoded that's >= a span's End.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17830)*
<!-- Sent from Reviewable.io -->
